### PR TITLE
Medcat refresh

### LINKED ISSUE #503: Simplify medcat
### Description of feature

I think that we over engineered it. A lot.

- [ ] Look at how MedCAT currently works outside of ehrapy
- [ ] Look at how we implemented it in ehrapy
- [ ] Remove the custom class MedCATsomething object in ehrapy
- [ ] ehrapy should just have as few functions as possible that take an AnnData object and maybe a MedCAT object that then annotate the text and store something in an AnnData slot. Maybe we have our own object but nothing that builds on top of the AnnData object like it does now.
- [ ] Then plot the stuff in the AnnData slot with ehrapy functions
- [ ] Implement everything.
- [ ] Add tests
- [ ] Update the tutorials

### PR DESCRIPTION
**Pull request by eroell**
**PR Checklist**

-   [x] This comment contains a description of changes (with reason)
-   [x] Referenced issue is linked
-   [x] If you've fixed a bug or added code that should be tested, add tests!
-   [x] Documentation in `docs` is updated

**Can I look at examples of how to use this version?**
Yes, in the parallel PR for the ehrapy tutorial here(to be inserted).

**Description of changes**
Resolves #503.

The goal is to simplify the use of medcat. In #503, the following TODOs have been outlined.

> - [x]  Look at how MedCAT currently works outside of ehrapy
> - [x] Look at how we implemented it in ehrapy
> - [x] Remove the custom class MedCATsomething object in ehrapy
> - [x] ehrapy should just have as few functions as possible that take an AnnData object and maybe a MedCAT object that then annotate the text and store something in an AnnData slot. Maybe we have our own object but nothing that builds on top of the AnnData object like it does now. _This first draft works without a MedCAT object. Something like this might be reintroduced if it enhances user experience and/or maintenance._
> - [x] Then plot the stuff in the AnnData slot with ehrapy functions _This first draft works by requiring to use `add_medcat_annotation_to_obs` before plotting, simplifying the overhead._
> - [x] Implement everything.
> - [x] Add tests
> - [ ] Update the tutorials

This draft removes the `MedCAT` and `EhrapyMedcat` classes and temporary moving-annotations-to-plotting.
Instead, it introduces scanpy-like methods
- `ep.tl.annotate_text`
- `ep.tl.get_medcat_annotation_overview`
- `ep.tl.add_medcat_annotation_to_obs`

**Technical details**

This does not move the `CAT` instance out of sight from the user: e.g. vanilla tutorials for MedCAT as well as sophisticated user adjustment can be done all on the `CAT` instance, nothing special here.

The desired `CAT` can then be passed to `ep.tl.annotate_text`, and ehrapy takes care of shaping the output into the right fields in the right format: such that `ep.tl.add_medcat_annotation_to_obs` can be used, followed by any plotting function which can use `.obs` entries.

**Questions**
- What do you think?
- Better suggestions for naming?
- Should the `run_unsupervised_training` convenience method be re-enabled?
- Should temporary move-to-obs be re-enabled?

**Todos if path is outlined**
- Add tests
- Add examples into docstrings

### CODE REVIEW

**Review by Zethson — COMMENTED** [review]
I think that this looks much better. It's going back to a design that we once had but then killed in favor of what we currently have.

1. I saw your PR on the tutorials repository. I assume that you ran it?
2. https://github.com/theislab/ehrapy/issues/591 - are we resolving this issue with this? Basically, can we make it easier for people to download the models? Or is that MedCat's problem?
3. Are there any new MedCat features that we could make use of? It's under active development and I only looked at it one year ago.
4. Related to 3 - can we get hierarchies somehow?
5. Is there a reasonable way for us to offer transferring the annotations into some kind of word2vec that people can append to a layer? Overkill?

Thank you so much!

**Comment by eroell** [comment]
> 1. I saw your PR on the tutorials repository. I assume that you ran it?

Yes, I ran it, using this branch of ehrapy: it used a stored `medmen_wstatus_2021_oct.zip` though, which is why it worked with an outdated link. This link has now been fixed there.

> 2. https://github.com/theislab/ehrapy/issues/591 - are we resolving this issue with this? Basically, can we make it easier for people to download the models? Or is that MedCat's problem?

This was an outdated link, which has been updated in the tutorial now.

> 3. Are there any new MedCat features that we could make use of? It's under active development and I only looked at it one year ago.

Did not check this in more detail yet.

> 4. Related to 3 - can we get hierarchies somehow?

Did not check this in more detail yet.

> 5. Is there a reasonable way for us to offer transferring the annotations into some kind of word2vec that people can append to a layer? Overkill?

Did not check this in more detail yet. Think this would be a very nice future addition, probably not doable in a clean way for the next release though?

File: ehrapy/tools/nlp/_medcat.py [hunk 1]
```diff
@@ -1,394 +1,254 @@
...
-    from anndata import AnnData
+    except ModuleNotFoundError:
+        pass
 
 
-class MedCAT:
-    """Wrapper class for Medcat. This class will hold references to the current AnnData object, which holds the data, the current model (with vocab and concept database) and should be
-    passed to all functions exposed to the ehrapy nlp API when required.
-    """
+def _filter_null_values(df: pd.DataFrame, column: str) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (eroell) [reply ehrapy/tools/nlp/_medcat.py:20]: Yes, just a relict from the previous implementation. Will remove

File: ehrapy/tools/nlp/_medcat.py [hunk 2]
```diff
@@ -1,394 +1,254 @@
...
+                    if entity_key in ["pretty_name", "cui", "type_ids", "types"]:
+                        single_entity[entity_key] = entities[entity_id][entity_key]
+                    elif entity_key == "meta_anns":
+                        single_entity[entity_key] = entities[entity_id][entity_key]["Status"]["value"]
+                flattened_annotated_dict[entry_nr] = single_entity
+                entry_nr += 1
+    return flattened_annotated_dict
+
+
+def _annotated_results_to_df(flattened_results: dict) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (eroell) [reply ehrapy/tools/nlp/_medcat.py:65]: Relict from earlier code - removed

File: ehrapy/tools/nlp/_medcat.py [hunk 3]
```diff
@@ -1,394 +1,254 @@
...
+def annotate_text(
+    adata: AnnData,
+    cat: CAT,
+    text_column: str,
+    key_added: str = "medcat_annotations",
+    n_proc: int = 2,
+    batch_size_chars: int = 500000,
+    copy: bool = False,
+) -> AnnData | None:
+    """Annotate the original free text data. Note this will only annotate non null rows.
```
[the remark this answers is not in the document]
Comment (eroell) [reply ehrapy/tools/nlp/_medcat.py:83]: Overlooked that here, thanks

File: ehrapy/tools/nlp/_medcat.py [hunk 4]
```diff
@@ -1,394 +1,254 @@
...
+    The result will be a DataFrame. It will be set as the annotated_results attribute for the passed MedCat object.
```
[the remark this answers is not in the document]
Comment (eroell) [reply ehrapy/tools/nlp/_medcat.py:84]: Nice thanks

File: ehrapy/tools/nlp/_medcat.py [hunk 5]
```diff
@@ -1,394 +1,254 @@
...
-        As an example:
-        Setting tuis=["T047", "T048"] will only annotate concepts (identified by a CUI (concept unique identifier)) in UMLS that are either diseases or
-        syndroms (T047) or mental/behavioural dysfunctions (T048).
+    """
+    if copy:
+        adata = adata.copy()
 
-        Args:
-            tuis: list of TUI's (default is
+    non_null_text = _filter_null_values(adata.obs, text_column)
```
[the remark this answers is not in the document]
Comment (eroell) [reply ehrapy/tools/nlp/_medcat.py:106]: Yes, now with some of your comments used its a bit more and I think indeed still readable

File: ehrapy/tools/nlp/_medcat.py [hunk 6]
```diff
@@ -1,394 +1,254 @@
...
-        # add column to additional to remove it later on
-        if add_cols is not None:
-            add_cols.append(name)
-        adata.obs[name] = (
-            df.groupby("row_nr").agg({"pretty_name": (lambda x: int(any(x.isin([name]))))}).astype("category")
+    # check whether the name is in the extracted entities to inform about possible typos
+    # currently, only the pretty_name column is supported
+    for i, nm in enumerate(name):
+        adata.obs[added_colname[i]] = (
+            df.groupby("row_nr").agg({"pretty_name": (lambda x, nm=nm: int(any(x.isin([nm]))))}).astype("category")
```
[the remark this answers is not in the document]
Comment (eroell) [reply ehrapy/tools/nlp/_medcat.py:246]: By the way what is your take here on using "yes" and "no", and not booleans?

File: ehrapy/tools/nlp/_medcat.py [hunk 7]
```diff
@@ -1,394 +1,254 @@
...
-
-    def update_cat(self, vocabulary: Vocab = None, concept_db: CDB = None):
-        """Updates the current MedCAT instance with new Vocabularies and Concept Databases.
 
-        Args:
-            vocabulary: Vocabulary to update to.
-            concept_db: Concept Database to update to.
-        """
-        self.cat = CAT(cdb=concept_db, config=concept_db.config, vocab=vocabulary)
+def _format_df_column(df: pd.DataFrame, column_name: str) -> list[tuple[int, str]]:
```
[the remark this answers is not in the document]
Comment (Zethson) [reply ehrapy/tools/nlp/_medcat.py:20]: I'd happily make this your call. Whatever you think is best

File: ehrapy/tools/nlp/_medcat.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zethson) [reply ehrapy/tools/nlp/_medcat.py]: Hmm. I prefer booleans I think

**Pull request merged**